File: C++/task#5.cpp
#include <iostream>
#include <vector>

using namespace std;

int powmod(int a, int x, int p) {
    int res = 1;
    for (int i = 0; i < x; i++) {
        res = res * a;
        res = res % p;
    }
    return res;
}

vector<int> cycle(int a) {
    vector<int> ost;
    int res = 1;

    while (true) {
        res = (res * a) % 10;

        // Проверяем, есть ли уже res в ost
        bool found = false;
        for (int val : ost) {
            if (val == res) {
                found = true;
                break;
            }
        }

        if (found) {
            // Цикл найден, прерываем цикл
            break;
        }

        ost.push_back(res);
    }

    return ost;
}

int main() {
    int a, b, c;

    cout << "Введите a, b, c: ";
    cin >> a >> b >> c;
    
    vector<int> res = cycle(a);

    int temp = powmod(b, c, res.size());

    int last;

    if (temp == 0) {
        last = res.back();
    } else {
        last = res[temp - 1];
    }

    cout << "Последняя цифра 'трёхэтажного числа' = "<< last << endl;
}

File: Python/task#5.py
def powmod(a, x, p):
    res = 1
    for _ in range(x):
        res = (res * a) % p
    return res

def cycle(a):
    ost = []
    res = 1
    while True:
        res = (res * a) % 10
        if res in ost:
            break
        ost.append(res)
    return ost

def main():
    a = int(input("Введите a: "))
    b = int(input("Введите b: "))
    c = int(input("Введите c: "))

    res = cycle(a)
    temp = powmod(b, c, len(res))

    if temp == 0:
        last = res[-1]
    else:
        last = res[temp - 1]

    print("Последняя цифра 'трёхэтажного числа' =", last)

if __name__ == "__main__":
    main()